db/db_impl.cc

```diff
@@ -2195,20 +2200,20 @@ autovector<Status> DBImpl::CreateColumnFamilyImpl(
         continue;
       }
 
-      edit_vec.emplace_back();
-      VersionEdit& edit = edit_vec.back().edit;
-      edit.AddColumnFamily(*column_family_name[i]);
+      VersionEdit* const edit_ptr = new VersionEdit();
+      edit_ptr->AddColumnFamily(*column_family_name[i]);
       uint32_t new_id =
           versions_->GetColumnFamilySet()->GetNextColumnFamilyID();
-      edit.SetColumnFamily(new_id);
-      edit.SetLogNumber(logfile_number_);
-      edit.SetComparatorName(cf_options[i]->comparator->Name());
-      edit_vec[i].mopt = MutableCFOptions(*cf_options[i], env_);
+      edit_ptr->SetColumnFamily(new_id);
+      edit_ptr->SetLogNumber(logfile_number_);
+      edit_ptr->SetComparatorName(cf_options[i]->comparator->Name());
+      const MutableCFOptions* const mopt_ptr =
+          new MutableCFOptions(*cf_options[i], env_);
 
       cfds.emplace_back(nullptr);
-      mutable_cf_options_list.emplace_back(&edit_vec.back().mopt);
```

Comment:
It is true that this line would place pointers to heap variables in another vector. However, when this function returns,  destructor of `edit_list` is called, which cleans up memory for several `autovector<VersionEdit*>`s, and ultimately cleans up memory for `VersionEdit*`. Note that in C++, dropping a pointer won't call `delete` of the content. e.g.

```
{
  std::vector<int*> x;
  {
    int a = 1;
    x.push_back(&a);
  }
    x.push_back(new int(1));
}
```

This code should run without any error, except `new int(1)` is leaked. So does this part of code.

Reply:
In your code snippet, you construct `a` inside the block and store its pointer to `x`. After the block (outside the block), `a` is destructed. `x` still has that address, which points to 👻  (UB). You should see errors if you try to use `*x[0]` (e.g., `std::cout << *x[0] << std::endl;`) after the block with ASAN check (note valgrind might not catch this). Reference: https://github.com/google/sanitizers/wiki/AddressSanitizerExampleUseAfterScope

Reply:
Attach valgrind report - https://gist.github.com/v1siuol/a39e9ec79522763f46ce379a8773bb97

There are two files (quite verbose): `valgrind-before.txt` is the output before the change, and  `valgrind-after.txt` is the output after the change. They are quite verbose and get truncated (you can download them.)

As you can see from the `valgrind-after.txt`, we still have reachable blocks and errors, even though we can pass the test :(

Reply:
Sharing the process when I was debugging:
```
Thread 1 "column_family_t" hit Breakpoint 2, terarkdb::DBImpl::CreateColumnFamilyImpl (this=0x7ffff6312000, cf_options=...,
    column_family_name=..., handle=...) at <cwd>/db/db_impl.cc:2215
2215          ++ok_count;
(gdb) p *edit_lists[71][0]
$8 = {max_level_ = 0, comparator_ = "leveldb.BytewiseComparator", log_number_ = 3, prev_log_number_ = 0, next_file_number_ = 0,
  max_column_family_ = 0, min_log_number_to_keep_ = 0, last_sequence_ = 0, has_comparator_ = true, has_log_number_ = true,
  has_prev_log_number_ = false, has_next_file_number_ = false, has_last_sequence_ = false, has_max_column_family_ = false,
  has_min_log_number_to_keep_ = false, deleted_files_ = std::set with 0 elements, new_files_ = std::vector of length 0, capacity 0,
  apply_callback_vec_ = {num_stack_items_ = 0, values_ = {{callback = 0x0, args = 0x0}, {callback = 0x0, args = 0x0}},
    vect_ = std::vector of length 0, capacity 0}, column_family_ = 72, is_column_family_drop_ = false, is_column_family_add_ = true,
  column_family_name_ = "cf1-72", is_open_db_ = false, is_in_atomic_group_ = false, remaining_entries_ = 0}
(gdb) c
Continuing.

Thread 1 "column_family_t" hit Breakpoint 3, terarkdb::DBImpl::CreateColumnFamilyImpl (this=0x7ffff6312000, cf_options=...,
    column_family_name=..., handle=...) at <cwd>/db/db_impl.cc:2220
2220        if (ok_count > 0) {  // write thread
(gdb) p *edit_lists[71][0]
$9 = {max_level_ = 0, comparator_ = "leveldb.BytewiseComparator", log_number_ = 0, prev_log_number_ = 1,
  next_file_number_ = 140737323541376, max_column_family_ = 0, min_log_number_to_keep_ = 0, last_sequence_ = 0, has_comparator_ = true,
  has_log_number_ = true, has_prev_log_number_ = false, has_next_file_number_ = false, has_last_sequence_ = false,
  has_max_column_family_ = false, has_min_log_number_to_keep_ = false, deleted_files_ = std::set with 0 elements,
  new_files_ = std::vector of length 549755170085, capacity 0 = {<error reading variable>
```
Reference: https://github.com/bytedance/terarkdb/blob/d4509bd51038b3618dc043bd3b1e20dea2b2921d/db/db_impl.cc#L2215-L2220

The first access is in the block (db_impl.cc:2215), everything looks good.

But the second access is outside the block (db_impl.cc:2220), boom.

Reply:
> In your code snippet, you construct `a` inside the block and store its pointer to `x`. After the block (outside the block), `a` is destructed. `x` still has that address, which points to 👻 (UB). You should see errors if you try to access `*x[0]` (e.g., `std::cout << *x[0] << std::endl;`) after the block with ASAN check (note valgrind might not catch this). Reference: https://github.com/google/sanitizers/wiki/AddressSanitizerExampleUseAfterScope

Agree on this. When I was reviewing the code, I didn't see any thread reading from the vector. Now this looks reasonable to me.